lib/cog/pipeline.ex

```diff
@@ -0,0 +1,337 @@
+defmodule Cog.Pipeline do
+
+  @default_output_policy :adapter
+  @default_command_timeout 60000
+
+  alias Carrier.Messaging.{ConnectionSup, Connection}
+  alias Cog.Events.PipelineEvent
+  alias Cog.Chat.Adapter, as: ChatAdapter
+  alias Cog.Pipeline.Destination
+  alias Cog.Command.Service.Tokens
+  alias Cog.Messages.ProviderRequest
+  alias Cog.Models.EctoJson
+  alias Cog.Pipeline.PrepareError
+  alias Cog.Pipeline.{CommandResolver, DataSignal, DoneSignal, ExecutionStageSup, ErrorSinkSup,
+                      InitialContext, InitialContextSup, OutputSinkSup, PermissionsCache}
+  alias Cog.Queries.User, as: UserQueries
+  alias Cog.Repo
+  alias Piper.Command.Ast
+  alias Piper.Command.Parser
+  alias Piper.Command.ParserOptions
+
+  use GenServer
+
+  require Logger
+
+  @type output_policy :: :adapter | :owner | :adapter_owner
+
+  defstruct [:policy, :owner, :request, :stages, :token, :conn, :started, :status]
+
+  @doc """
+  Starts a new pipeline and prepares it to begin
+  processing
+
+  ## Options
+  * `:request` - Originating `Cog.Messages.ProviderRequest`. Required.
+  * `:owner` - Pid of pipeline owner. Optional.
+  * `:output_policy` - Controls where to send output. Valid values for this option are:
+    * `:adapter` - Pipeline will send output via `Cog.Chat.Adapter` (default)
+    * `:owner` - Pipeline will forward output to the owning process. Requires `:owner` option.
+    * `:adapter_owner` - Pipeline sends output via `Cog.Chat.Adapter` _and_ forwards
+       a copy to the owning process. Requires `:owner` option.
+  """
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, [opts])
+  end
+
+  @doc """
+  Tells a prepared pipeline to begin executing
+  """
+  def run(pipeline) do
+    GenServer.call(pipeline, :run, :infinity)
+  end
+
+  @doc """
+  Notifies pipeline when processing is finished
+  """
+  def notify(pipeline) do
```

Comment:
"notify" seems like a general concept, but "teardown" is specific. Should this function be renamed to "teardown" instead?